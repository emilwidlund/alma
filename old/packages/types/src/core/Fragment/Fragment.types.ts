import { z } from 'zod';

import { FragmentSchema } from './Fragment';

export type Fragment = z.infer<typeof FragmentSchema>;
