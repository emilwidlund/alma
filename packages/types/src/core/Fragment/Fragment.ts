import { z } from 'zod';

export const FragmentSchema = z.string();
