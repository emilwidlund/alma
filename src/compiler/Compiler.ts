import { assign, defMain, FLOAT0, FLOAT1, Sym, vec4 } from "@thi.ng/shader-ast";
import { GLSLVersion } from "@thi.ng/shader-ast-glsl";
import { ShaderFn, shaderSourceFromAST, ShaderSpec, ShaderUniformSpecs } from "@thi.ng/webgl";
import { Root } from "../nodes/common/Root/Root";

const createShaderSpec = (
    root: ShaderFn,
    uniforms?: ShaderUniformSpecs
): ShaderSpec => {
    return {
        vs: (gl, _, ins, outs) => [
            defMain(() => [assign(outs.vUv, ins.uv), assign(gl.gl_Position, vec4(ins.position, FLOAT0, FLOAT1))])
        ],
        fs: root,
        uniforms,
        attribs: { position: 'vec2', uv: 'vec2' },
        varying: { vUv: 'vec2' }
    };
};

export const compile = <TUniforms extends ShaderUniformSpecs>(root: Root, uniforms: TUniforms, target: GLSLVersion) => {
    const shaderSpec = createShaderSpec(root.resolve.bind(root), uniforms);
    
    return shaderSourceFromAST(shaderSpec, 'fs', target);
}