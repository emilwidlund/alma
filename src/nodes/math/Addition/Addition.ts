import { Output } from '@nodl/core';
import { add } from '@thi.ng/shader-ast';
import { combineLatest, map } from 'rxjs';

import { PrimSchema } from '../../../schemas/Prim/Prim';
import { ABPrimNode } from '../../internal/ABPrimNode/ABPrimNode';

export class Addition extends ABPrimNode {
    name = 'Addition';

    outputs = {
        output: new Output({
            name: 'Output',
            type: PrimSchema,
            observable: combineLatest([this.inputs.a, this.inputs.b]).pipe(map(inputs => add(...inputs)))
        })
    };
}
