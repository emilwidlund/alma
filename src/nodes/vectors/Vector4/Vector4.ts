import { Input, Node, Output } from '@nodl/core';
import { float, Prim, Term, Vec, vec3, vec4 } from '@thi.ng/shader-ast';
import { combineLatest, map } from 'rxjs';

import { FloatSchema } from '../../../schemas/Float/Float';
import { Vec4Schema } from '../../../schemas/Vec4/Vec4';

export class Vector4 extends Node {
    name = 'Vector 4';

    inputs = {
        x: new Input({
            name: 'X',
            type: FloatSchema,
            defaultValue: float(0)
        }),
        y: new Input({
            name: 'Y',
            type: FloatSchema,
            defaultValue: float(0)
        }),
        z: new Input({
            name: 'Z',
            type: FloatSchema,
            defaultValue: float(0)
        }),
        w: new Input({
            name: 'W',
            type: FloatSchema,
            defaultValue: float(1)
        })
    }
    
    outputs = {
        output: new Output({
            name: 'Output',
            type: Vec4Schema,
            observable: combineLatest([this.inputs.x, this.inputs.y, this.inputs.z, this.inputs.w]).pipe(map(inputs => vec4(...inputs)))
        })
    };
}
