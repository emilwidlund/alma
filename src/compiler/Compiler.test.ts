import { GLSLVersion } from '@thi.ng/shader-ast-glsl';
import {Fragment} from '../nodes/core/Fragment/Fragment';
import {Vector4} from '../nodes/vectors/Vector4/Vector4';
import {Addition} from '../nodes/math/Addition/Addition';
import {compile} from './Compiler';
import { float } from '@thi.ng/shader-ast';

describe('Compiler', () => {
    it('should produce valid GLSL code', () => {
        const fragmentNode = new Fragment();
        const compiledOutput = compile(fragmentNode, {}, GLSLVersion.GLES_300);

        expect(compiledOutput).toMatchSnapshot();
    });

    it('should produce valid GLSL code from a complex graph', () => {
        const a = new Vector4();
        const b = new Vector4();
        const add = new Addition();
        const fragmentNode = new Fragment();

        a.inputs.x.next(float(.3));
        a.inputs.y.next(float(.5));
        a.inputs.z.next(float(.1));

        b.inputs.x.next(float(1));
        b.inputs.y.next(float(.02));
        b.inputs.z.next(float(.6));
        b.inputs.w.next(float(1));

        a.outputs.output.connect(add.inputs.a);
        b.outputs.output.connect(add.inputs.b);
        add.outputs.output.connect(fragmentNode.inputs.color);

        const compiledOutput = compile(fragmentNode, {}, GLSLVersion.GLES_300);

        expect(compiledOutput).toMatchSnapshot();
    });
})