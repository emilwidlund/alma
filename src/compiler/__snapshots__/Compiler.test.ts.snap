// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Compiler should produce valid GLSL code 1`] = `
"#version 300 es
#ifdef GL_FRAGMENT_PRECISION_HIGH
precision highp int;
precision highp float;
#else
precision mediump int;
precision mediump float;
#endif
#ifndef PI
#define PI 3.141592653589793
#endif
#ifndef TAU
#define TAU 6.283185307179586
#endif
#ifndef HALF_PI
#define HALF_PI 1.570796326794896
#endif

in vec2 vUv;
layout(location=0) out vec4 fragColor;
void main() {
fragColor = vec4(0.0, 0.0, 0.0, 1.0);
}"
`;

exports[`Compiler should produce valid GLSL code from a complex graph 1`] = `
"#version 300 es
#ifdef GL_FRAGMENT_PRECISION_HIGH
precision highp int;
precision highp float;
#else
precision mediump int;
precision mediump float;
#endif
#ifndef PI
#define PI 3.141592653589793
#endif
#ifndef TAU
#define TAU 6.283185307179586
#endif
#ifndef HALF_PI
#define HALF_PI 1.570796326794896
#endif

in vec2 vUv;
layout(location=0) out vec4 fragColor;
void main() {
fragColor = (vec4(0.3, 0.5, 0.1, 1.0) + vec4(1.0, 0.02, 0.6, 1.0));
}"
`;
